### DISCUSSION TITLE
2021 Miller User Survey results

### DISCUSSION THREAD

**Discussion by johnkerl** [johnkerl/miller#670] [Ideas]
There were about 40 responses to the [2021 Miller User Survey](https://github.com/johnkerl/miller/discussions/542). It's deeply gratifying to get such thoughtful, useful feedback! I thought I'd share some of the findings.

# Use-cases

These are all over the map, and pleasantly so. All along I intended Miller to be multi-purpose -- it seems many of us are solving the same problems, sometimes ad-hoc, and there was a people-empowering niche to fill. The results here confirm this. One of my favorite quotes ever was @eddelbuettel referring to Miller as an _underappreciated swiss-army chainsaw_.

<img width="775" alt="Screen Shot 2021-09-19 at 9 40 11 AM" src="https://user-images.githubusercontent.com/2008794/133930041-191d636e-6d67-4fd2-87c5-577769477e29.png">

<img width="776" alt="Screen Shot 2021-09-19 at 9 40 39 AM" src="https://user-images.githubusercontent.com/2008794/133930049-99b83ca5-d34d-40f5-9d3e-9e75fb2b3965.png">

# File formats

Here too the results are all over the map -- but with a marked emphasis on CSV. When I originally wrote Miller, back in 2015, [DKVP](https://johnkerl.org/miller6/file-formats/#dkvp-key-value-pairs) was the format I most often used (although I use primarily CSV, TSV, PPRINT, XTAB, and JSON now); DKVP was (and is) Miller's default file format. However, most people value CSV -- while each of Miller's supported formats is useful to some survey respondents, _all_ respondents flagged CSV.

<img width="768" alt="Screen Shot 2021-09-19 at 9 40 52 AM" src="https://user-images.githubusercontent.com/2008794/133930265-5cc87ff0-22ca-4875-a6e6-e12c34c8d70a.png">

As a result of this feedback, the [Miller 6 docs](https://johnkerl.org/miller6) are more CSV-centric. Up-front examples that used the `abixy` sample DKVP file in the Miller-5 documents now use the `example.csv` file, and many examples now use CSV and JSON.

# Documents

The most-common response was that the docs are just right. But a close second response was that it's too hard for people to find what they're looking for _within_ that material.

Some free-form comments:

* There is a lot of material, but it's better to have too much than too little
* An ask for an acronyms dictionary
* More examples
* The docs give too much focus, too soon, on advanced material and too little focus on simpler (and more common) use-cases
* Most-important pages (such as the verb-list) were buried deep in, and should be a click or two away from the landing page

<img width="775" alt="Screen Shot 2021-09-19 at 9 41 43 AM" src="https://user-images.githubusercontent.com/2008794/133930312-038f3751-de98-4f68-b7db-1c06855c1629.png">

As a result of this feedback, the [Miller 6 documents](https://johnkerl.org/miller6) have been completely reworked relative to the [Miller 5 documents](https://miller.readthedocs.io).

<img width="1248" alt="Screen Shot 2021-09-19 at 10 08 51 AM" src="https://user-images.githubusercontent.com/2008794/133930744-5b7b9853-107f-45a4-a793-dd910e057a86.png">

* The site-generator has been ported from Sphinx to Mkdocs, with improved look and feel. As a bonus, the new Mkdocs structure is much more mobile-friendly, working well on desktop as well as mobile.
* The left-hand table of contents is always in view (a finger-tap away on mobile), showing current context within the doc site.
* Likewise, the in-page table of contents on the right-hand side allows for quick navigation for sections of the current page.
* The _Search_ feature is at the top right of every page.
* _Quick links_ is at the top center of every page, allowing repeat visitors quick access to most-used reference information.
* The new [glossary](https://johnkerl.org/miller6/glossary) is a particularly useful addition, thanks to an idea from @NikosAlexandris.
* Code snippets now have their command-line inputs on a darker background, with bold text -- thanks to an idea from @aaronwolen.
* The site structure is now multi-part: _Getting started_ with extended narrative and introductory examples; _Miller in more detail_ for next-level / early-on reading; _FAQs and examples_ with non-theoretical deep-diving into specific use-cases; and lastly thorough reference sections for those times when people need the most precise information.
* The reference pages have been split up into more, but shorter, pages.
* In addition to the survey, more doc-related feedback is on GitHub Discussions at https://github.com/johnkerl/miller/discussions/621. @aborruso and @NikosAlexandris have contributed fantastic review and discussion.

# Output colorization

In some recent covert art at https://github.com/johnkerl/miller I intended only to show visually that records, as lists of key-value pairs, are the common structure underlying Miller's handling of seemingly distinct file formats. This is, in fact, what makes it possible for Miller to be a swiss-army chainsaw. But one respondent was dismayed that this coloring was not a live feature.

<img width="497" alt="Screen Shot 2021-09-19 at 10 24 30 AM" src="https://user-images.githubusercontent.com/2008794/133931159-cbf6affa-288c-47bb-9e15-5daf25d1a36d.png">

I'd never thought of that, but realized it was not only doable but desirable -- and after all, `jq` does something similar with JSON files.

As a result, in [Miller 6](https://johnkerl.org/miller6/new-in-miller-6/) we now have [output colorization](https://johnkerl.org/miller6/output-colorization) -- what was once cover art is now something we can use interactively.

# Community

As a clear growth opportunity, most respondents were not aware of [current Miller community options](https://johnkerl.org/miller6/community).

<img width="776" alt="Screen Shot 2021-09-19 at 9 41 51 AM" src="https://user-images.githubusercontent.com/2008794/133931361-fabdcd34-a55f-441e-b9a3-a80f0127cd25.png">

# What should be next?

Development on Miller 6 continues unabated.

People were most interested in testing:

<img width="775" alt="Screen Shot 2021-09-19 at 9 42 05 AM" src="https://user-images.githubusercontent.com/2008794/133931444-5651c916-15c3-4508-92f6-c9edc01b6aa7.png">

I hope to have a beta release of Miller 6 out soon for general evaluation.

As always, please keep the feedback coming! Six years ago, Miller was a personal project; good feedback is helping it to become much more than I had ever anticipated.

* https://github.com/johnkerl/miller/discussions
* https://github.com/johnkerl/miller/issues

**Comment by aborruso** [comment]
Hi @johnkerl thank you very much, really really interesting.

I have a stupid question: why is the sum of the percentages is more than 100 (in more than one graph)?
![image](https://user-images.githubusercontent.com/30607/133934508-9a81d5cb-bde3-4f41-998a-a76b1cb6c1e9.png)

**Reply by eddelbuettel** [reply]
Maybe because the answers were non-exclusive?

**Reply by aborruso** [reply]
> Maybe because the answers were non-exclusive?

probably you are right, thank you

**Reply by johnkerl** [reply]
@aborruso @eddelbuettel indeed! :)
